### PR TITLE
Store VAAs by TX hash in bigtable

### PR DESCRIPTION
**Pull request by kev1n-peters**
Addresses #93

### CODE REVIEW

**Review by panoel — APPROVED** [review]

File: watcher/src/databases/types.ts [hunk 1]
```diff
@@ -15,6 +15,16 @@ export type BigtableMessagesRow = {
     };
   };
 };
+export interface BigtableVAAsByTxHashRow {
+  key: string;
+  data: {
+    // column family
+    info: {
+      // columns
+      vaaKeys: { value: string[]; timestamp: string };
```
Comment (kev1n-peters) [inline watcher/src/databases/types.ts:24]: I don't think bigtable supports array data types natively. Maybe we can delimit the `vaaKeys` and store the result as `value: string`? @evan-gray

Comment (kev1n-peters) [reply watcher/src/databases/types.ts:24]: Another option would be to store the VAA keys as a JSON object or array, e.g.:
```
// object
{ vaaKeys: ["vaa1", ... , "vaaN"] }

// array
["vaa1", ... , "vaaN"]
```

Comment (kev1n-peters) [reply watcher/src/databases/types.ts:24]: I guess I like storing the keys as CSVs, because it's the simplest solution

Comment (kev1n-peters) [reply watcher/src/databases/types.ts:24]: Agreed on this: "JSON.stringify on the way in and JSON.parse on the way out seem right to me"

File: watcher/src/databases/BigtableDatabase.ts [hunk 2]
```diff
@@ -111,9 +122,22 @@ export class BigtableDatabase extends Database {
...
+      ([txHashRowKey, vaaRowKeys]) => ({
+        key: txHashRowKey,
+        data: {
+          info: {
+            vaaKeys: { value: JSON.stringify(vaaRowKeys), timestamp: '0' },
+          },
+        },
+      })
+    );
+    await Promise.all([table.insert(rowsToInsert), vaasByTxHashTable.insert(txHashRowsToInsert)]);
```
Comment (evan-gray) [inline watcher/src/databases/BigtableDatabase.ts:140]: Smooth.

**Review by evan-gray — APPROVED** [review]
Approving since I believe the core monitor code works, but left a couple comments if you want to tidy them up first.

File: cloud_functions/src/getVaasByTxHash.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      res.status(404);
+      res.json({ error: 'no signed VAAs for tx' });
+      return;
+    }
+    const result = signedVAAs[0].map((row) => ({
+      // gross 🤮
+      id: (row.id as unknown as Buffer).toString('utf8'),
+      vaaBytes: row.data.info.bytes[0].value.toString('hex'),
+    }));
+    res.json(result);
```
[the remark this answers is not in the document]
Comment (kev1n-peters) [reply cloud_functions/src/getVaasByTxHash.ts:45]: I agree, changed

File: README.md [hunk 4]
```diff
@@ -44,7 +44,7 @@ Currently three options to load and save data:
 3. google bigtable with firestore:
    > you will need to set up your credentials: https://cloud.google.com/docs/authentication/provide-credentials-adc and set GOOGLE_APPLICATION_CREDENTIALS to the path of your credentials
    > set up the instance and table: https://cloud.google.com/bigtable/docs/creating-instance
-   > set DB_SOURCE="bigtable", BIGTABLE_INSTANCE_ID, and BIGTABLE_TABLE_ID
+   > set DB_SOURCE="bigtable", BIGTABLE_INSTANCE_ID, BIGTABLE_TABLE_ID, and BIGTABLE_VAAS_BY_TX_HASH_TABLE_ID
```
[the remark this answers is not in the document]
Comment (kev1n-peters) [reply README.md]: fixed

File: watcher/scripts/backfillVAAsByTxHash.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      throw new Error('bigtable is undefined');
+    }
+    const instance = bt.bigtable.instance(bt.instanceId);
+    const messageTable = instance.table(bt.tableId);
+    const vaasByTxHashTable = instance.table(bt.vaasByTxHashTableId);
+
+    let log = ora(`Reading rows from ${bt.tableId}...`).start();
+    const observedMessages = await messageTable.getRows(); // TODO: pagination
+    const vaasByTxHash: { [key: string]: string[] } = {};
+    observedMessages[0].forEach((msg) => {
```
[the remark this answers is not in the document]
Comment (kev1n-peters) [reply watcher/scripts/backfillVAAsByTxHash.ts]: fixed

File: watcher/scripts/backfillVAAsByTxHash.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      const txHash = msg.data.info.txHash[0].value;
+      const splitId = msg.id.split('/');
```
[the remark this answers is not in the document]
Comment (kev1n-peters) [reply watcher/scripts/backfillVAAsByTxHash.ts]: fixed

File: watcher/scripts/backfillVAAsByTxHash.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        key: txHashRowKey,
+        data: {
+          info: {
+            vaaKeys: { value: JSON.stringify(vaaRowKeys), timestamp: '0' },
+          },
+        },
+      })
+    );
+    let numWritten = 0;
+    while (numWritten < rowsToInsert.length) {
```
[the remark this answers is not in the document]
Comment (kev1n-peters) [reply watcher/scripts/backfillVAAsByTxHash.ts]: I decided to at least use `readChunks` that you wrote instead of rolling my own here

**Review by evan-gray — APPROVED** [review]

**Pull request merged**